client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,144 @@
+import { getVariant } from '../data/gameData';
+import * as deckRules from '../rules/deck';
+import { GameAction } from '../types/actions';
+import CardState from '../types/CardState';
+import GameMetadata from '../types/GameMetadata';
+import Variant from '../types/Variant';
+
+const cardDeductionReducer = (
+  deck: readonly CardState[],
+  action: GameAction,
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+) => {
+  switch (action.type) {
+    case 'cardIdentity':
+    case 'clue':
+    case 'discard':
+    case 'play':
+    case 'draw': {
+      return makeDeductions(deck, hands, metadata);
+    }
+    default: {
+      return Array.from(deck);
+    }
+  }
+};
+
+export default cardDeductionReducer;
+
+const makeDeductions = (
+  deck: readonly CardState[],
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+) => {
+  const newDeck: CardState[] = Array.from(deck);
+  const variant = getVariant(metadata.options.variantName);
+  const cardCountMap = getCardCountMap(variant);
+
+  // if our variant shows played/discarded cards, then they're known by everyone
+  newDeck.filter((card) => !hands.some((hand) => hand.includes(card.order)))
+    .filter((card) => card.suitIndex !== null && card.rank !== null)
+    .forEach((card) => {
+      cardCountMap[card.suitIndex!][card.rank!] -= 1;
+    });
```

Comment:
This can be simplified to use `card.location`. Also, it's expensive to use `filter` when you're just going to iterate over it's result.
```suggestion
  // if our variant shows played/discarded cards, then they're known by everyone
  for (const card of newDeck) {
    if ((card.location === 'discard' || card.location === 'playStack') && card.SuitIndex !== null && card.rank !== null) {
      cardCountMap[card.suitIndex][card.rank] -= 1;
    }
  }
```